host/docs/uhd_cal_tx_dc_offset.1
.TH "uhd_cal_tx_dc_offset" "1" "3.7.0" "UHD" "User Commands"
.SH NAME
uhd_cal_tx_dc_offset \- Generate TX DC Offset Calibration Table for a UHD Device
.SH DESCRIPTION
The Universal Software Radio Peripheral Hardware Drivers handle calibration
information. This application reduces TX DC offset within the specified frequency
range using transmit leakage into the receive path. Remove all external hardware
from the RF antenna ports before using this application. This can take up to
several minutes with default settings.
.LP
NOTE: This application can only be used with the WBX, SBX, XCVR2450, and RFX daughterboards.
.SH OPTIONS
.TP
\fB\-\-help\fR
help message
.TP
\fB\-\-verbose\fR
enable some verbose
.TP
\fB\-\-args\fR arg
device address args [default = ""]
.TP
\fB\-\-tx_wave_freq\fR arg (=507123)
Transmit wave frequency in Hz
.TP
\fB\-\-tx_wave_ampl\fR arg (=0.69999999999999996)
Transmit wave amplitude in counts
.TP
\fB\-\-rx_offset\fR arg (=934400)
RX LO offset from the TX LO in Hz
.TP
\fB\-\-freq_start\fR arg
Frequency start in Hz (will use beginning of frequency range if unspecified)
.TP
\fB\-\-freq_stop\fR arg
Frequency stop in Hz (will use end of frequency range if unspecified)
.TP
\fB\-\-freq_step\fR arg (=7300000)
Step size for LO sweep in Hz
.TP
\fB\-\-nsamps\fR arg (=10000)
Samples per data capture
.PP
.SH SEE ALSO
UHD documentation:
.B http://files.ettus.com/manual/
.LP
GR-UHD documentation:
.B http://gnuradio.org/doc/doxygen/page_uhd.html
.LP
Other UHD programs:
.sp
uhd_cal_rx_iq_balance(1) uhd_cal_tx_iq_balance(1)
.SH AUTHOR
This manual page was written by Maitland Bottoms and Nicholas Corgan
for the Debian project (but may be used by others).
.SH COPYRIGHT
Copyright (c) 2012-2022 Ettus Research, A National Instruments Brand
.LP
This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.
.LP
This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.


CHANGELOG
Change Log for Releases
==============================

## 004.002.000.000
* b200
  - Re-sync times
  - Fix overflow handling
  - Move the B200 radio control core into usrp/b200/
* cal
  - Use safe version of set_thread_priority()
  - Fix handling of discontinuities in power calibration data
* chdr
  - Rename var max_size_bytes to avoid confusion
* ci
  - Weekly builds for UHD-4.2 docker images
  - Update builds to macOS 12 Monterey
  - Update commit vcpkg and CMake version
  - Set continue on error and reduce timeout to 60
  - Remove obsolete x4xx pipelines
  - Upload devtest logs as artifact
  - Propagate downloader errors and lower timeout
  - Enable ctest on macOS builds
  - Flash fpga on x410 ATS runs
  - Add mpm folder to triggers
  - Change yaml dependency to ruamel
  - Add support for Ubuntu 22.04
  - Remove Fedora 33 and add Fedora 35
  - Add libuhd python dependencies to macOS
  - Add x410 test to mono pipeline
  - Enable building of init_usrp
  - Make uhdSrcDir directly reference Build.SourcesDirectory
  - Build gnuradio and gr-ettus
  - Add gnuradio deps to images
  - Generate installed binaries for downstream use
  - Add displayNames to parameters
  - Add embedded builds to uhd mono pipeline
  - Separate pipeline builds for different os
  - Refactor dockerOSName to buildOSName
  - Build on macOS
  - Device wait to redlock scope for Vivado close
  - Add Fedora 34 and remove Fedora 32
  - rebuild docker images weekly
  - Refactor installers and add Windows support
  - Add gnuradio gr-ettus to oe builds
  - Update template to use checkout_meta_ettus
  - Add custom boost version support
  - Remove documentation-only changes from pipeline runs
  - Enable batch CI
  - Split CI and PR pipelines for mono pipeline
  - Enable custom CXX flags, enable -Werror
  - Let make keep building upon failure
  - Add clang as a compiler to all Fedora and Ubuntu containers
  - Add CLA assistant
* cmake
  - mpm: Skip installing binaries for SIM
  - doxygen: Make MATHJAX_RELPATH configurable via CMake
  - ncurses: fix building with split tinfo
  - Remove libatomic check on macOS
  - Added libatomic check for boost/lockfree/queue.hpp
  - uhdboost: Fix check for UHD_BOOST_REQUIRED being set
  - Replace distutils.sysconfig with sysconfig
  - Replace distutils with CMake for version checks
  - Set debug to -Og for Clang builds
  - tests: Conditionally compile tests for X400
  - Use LooseVersion to ensure correct version comparisons
  - Fix rfnoc-example (CMake paths)
  - Fix issues with static builds and CMRC
  - Replace CMAKE_{SOURCE,BINARY}_DIR with UHD_*_DIR
  - tests: Add build-python path to PYTHONPATH
  - Add check for libatomic linking requirement
  - Remove duplicate entry in LIBUHD_PYTHON_GEN_SOURCE
  - Fix VS names and use relative for images
  - remove redundant include
  - Correctly set and unset any CMAKE_REQUIRED variables
  - Fix finding PkgConfig to work robustly (without CMake warnings)
* config
  - Fix clang fallthrough syntax
* conversion
  - Saturate transmit IQ levels on NEON architectures.
* convert
  - Make narrowing conversions saturate
  - Add benchmarking abilities
  - Minor cleanup
* core
  - Remove boost::math in favor of std cmath
* dbsrx
  - Fix issue with loop variable
* debian
  - Update version number to 4.2.0
  - Update control and package building
* debs
  - Update upload_debs script
* deps
  - rpclib: Replace distutils.dir_util with shutil
* devtest
  - Clarify data type in multi_usrp_test::send_waveform()
  - Add receive stability test to B2xx devtest
  - Add receive stability test
* dissectors
  - Fix whitespace formatting in CMake files
  - Fix inclusion of glib.h and Python version
* docs
  - multi_usrp: Clarify GPIO source bank meaning
  - x4xx: Document configuring eth0 static IP
  - n310: Add Filter API section
  - e31x: Update information on GPIO pin header
  - Add 100GigE documentation
  - Move X4x0 GPIO API to be subpage
  - Update E320 docs
  - x4xx: Remove redundant GPIO section
  - rfnoc: Add doxygen tag for missing parameter
  - Improve table on 'identification'
  - Update instructions for changing hostname
  - Update sfp port config location
  - Update manual for new X410 default targets
  - Fix Doxygen warnings
  - Fix reference to RFNoC documentation
  - rdtesting: Remove invalid rate configuration
  - stream_args: Clarify usage of stream_args_t::channels
  - n3xx: Add info on customizable band edges and gain profiles
  - x4xx: Add new FPGA image descriptions
  - Improve documentation for replay block
  - Remove superfluous stylesheet
  - x310: Remove reference to ORC
  - Update n3xx tuning notes
  - Remove full path names from Doxygen generation
  - Add shim Sphinx config for readthedocs
  - Improve page on RFNoC block properties
  - Fix page on GPIO
  - Fix MathJax formulae rendering
  - Amend page on RFNoC properties
  - Remove bmaptool instructions for writing filesystems
  - Remove obsolete man pages
  - Update DPDK documentation
  - Improve documentation for properties and -propagation
  - Several minor manual improvements
  - x410: Document GPIO API and capabilities
  - Collect all RFNoC block controllers in a module in the manual
  - Align dependencies and bump deb package versions
  - Clarify set/get_gpio_attr() and GPIO banks
  - Fix GPIO documentation example
  - x410: Fix info on loading SD card images with bmaptool
  - Improve docs for rx_streamer::recv() on overruns
  - sync: Update page on synchronization
  - Fix typo in ZBX Block Diagram
  - usrp_x4xx: improve filesystem update instructions
  - x4xx: Update information on CPLD updating
* dpdk
  - Add support for DPDK 18.11 API
  - Correct MTU warning message
  - Support new MTU discovery
  - Disable warnings for using an experimental feature
  - Upgrade to DPDK 19.11 API
* e31x/e320
  - Amend LO-locked sensor names
* e320
  - mpm: Remove monitor thread
* e3xx
  - Fix frequency querying
  - Remove unused constant
* examples
  - gpio: Separate bank and port arguments

host/utils/uhd_cal_tx_dc_offset.cpp
//
// Copyright 2010,2012,2014 Ettus Research LLC
// Copyright 2018 Ettus Research, a National Instruments Company
//
// SPDX-License-Identifier: GPL-3.0-or-later
//

#include "usrp_cal_utils.hpp"
#include <uhd/usrp/multi_usrp.hpp>
#include <uhd/utils/algorithm.hpp>
#include <uhd/utils/paths.hpp>
#include <uhd/utils/safe_main.hpp>
#include <boost/format.hpp>
#include <boost/program_options.hpp>
#include <chrono>
#include <complex>
#include <ctime>
#include <functional>
#include <iostream>

namespace po = boost::program_options;

/***********************************************************************
 * Tune RX and TX routine
 **********************************************************************/
static double tune_rx_and_tx(
    uhd::usrp::multi_usrp::sptr usrp, const double tx_lo_freq, const double rx_offset)
{
    // tune the transmitter with no cordic
    uhd::tune_request_t tx_tune_req(tx_lo_freq);
    tx_tune_req.dsp_freq_policy = uhd::tune_request_t::POLICY_MANUAL;
    tx_tune_req.dsp_freq        = 0;
    usrp->set_tx_freq(tx_tune_req);

    // tune the receiver
    double rx_freq        = usrp->get_tx_freq() - rx_offset;
    double min_fe_rx_freq = usrp->get_fe_rx_freq_range().start();
    double max_fe_rx_freq = usrp->get_fe_rx_freq_range().stop();
    uhd::tune_request_t rx_tune_req(rx_freq);
    rx_tune_req.dsp_freq_policy = uhd::tune_request_t::POLICY_MANUAL;
    rx_tune_req.dsp_freq        = 0;
    if (rx_freq < min_fe_rx_freq)
        rx_tune_req.dsp_freq = rx_freq - min_fe_rx_freq;
    else if (rx_freq > max_fe_rx_freq)
        rx_tune_req.dsp_freq = rx_freq - max_fe_rx_freq;
    usrp->set_rx_freq(rx_tune_req);

    wait_for_lo_lock(usrp);

    return usrp->get_tx_freq();
}

/***********************************************************************
 * Main
 **********************************************************************/
int UHD_SAFE_MAIN(int argc, char* argv[])
{
    std::string args, subdev, serial;
    double tx_wave_freq, tx_wave_ampl, rx_offset;
    double freq_start, freq_stop, freq_step;
    size_t nsamps;
    double precision;

    po::options_description desc("Allowed options");
    // clang-format off
    desc.add_options()
        ("help", "help message")
        ("verbose", "enable some verbose")
        ("args", po::value<std::string>(&args)->default_value(""), "device address args [default = \"\"]")
        ("subdev", po::value<std::string>(&subdev), "Subdevice specification (default: first subdevice, often 'A')")
        ("tx_wave_freq", po::value<double>(&tx_wave_freq)->default_value(507.123e3), "Transmit wave frequency in Hz")
        ("tx_wave_ampl", po::value<double>(&tx_wave_ampl)->default_value(0.7), "Transmit wave amplitude")
        ("rx_offset", po::value<double>(&rx_offset)->default_value(.9344e6), "RX LO offset from the TX LO in Hz")
        ("freq_start", po::value<double>(&freq_start), "Frequency start in Hz (do not specify for default)")
        ("freq_stop", po::value<double>(&freq_stop), "Frequency stop in Hz (do not specify for default)")
        ("freq_step", po::value<double>(&freq_step)->default_value(default_freq_step), "Step size for LO sweep in Hz")
        ("nsamps", po::value<size_t>(&nsamps), "Samples per data capture")
        ("precision", po::value<double>(&precision)->default_value(default_precision), "Correction precision (default=0.0001)")
    ;
    // clang-format on

    po::variables_map vm;
    po::store(po::parse_command_line(argc, argv, desc), vm);
    po::notify(vm);

    // print the help message
    if (vm.count("help")) {
        std::cout << boost::format("USRP Generate TX DC Offset Calibration Table %s")
                         % desc
                  << std::endl;
        std::cout << "This application measures leakage between RX and TX on a "
                     "transceiver daughterboard to self-calibrate.\n"
                     "Note: Not all daughterboards support this feature. Refer to the "
                     "UHD manual for details.\n"
                  << std::endl;
        return EXIT_FAILURE;
    }

    // Create a USRP device
    uhd::usrp::multi_usrp::sptr usrp = setup_usrp_for_cal(args, subdev, serial);

    if (not vm.count("nsamps"))
        nsamps = size_t(usrp->get_rx_rate() / default_fft_bin_size);

    // create a receive streamer
    uhd::stream_args_t stream_args("fc32"); // complex floats
    uhd::rx_streamer::sptr rx_stream = usrp->get_rx_stream(stream_args);

    // create a transmit streamer
    uhd::tx_streamer::sptr tx_stream = usrp->get_tx_stream(stream_args);

    // create a transmitter thread
    std::atomic_flag transmit = ATOMIC_FLAG_INIT;
    transmit.test_and_set();
    auto transmitter = std::thread(
        std::bind(&tx_thread, &transmit, usrp, tx_stream, tx_wave_freq, tx_wave_ampl));

    // re-usable buffer for samples
    std::vector<samp_type> buff;

    // store the results here
    std::vector<result_t> results;

    if (not vm.count("freq_start"))
        freq_start = usrp->get_fe_tx_freq_range().start();
    if (not vm.count("freq_stop"))
        freq_stop = usrp->get_fe_tx_freq_range().stop();

    // check start and stop frequencies
    if (freq_start < usrp->get_fe_tx_freq_range().start()) {
        std::cerr << "freq_start must be " << usrp->get_fe_tx_freq_range().start()
                  << " or greater for this daughter board" << std::endl;
        return EXIT_FAILURE;
    }
    if (freq_stop > usrp->get_fe_tx_freq_range().stop()) {
        std::cerr << "freq_stop must be " << usrp->get_fe_tx_freq_range().stop()
                  << " or less for this daughter board" << std::endl;
        return EXIT_FAILURE;
    }

    // check rx_offset
    double min_rx_offset =
        usrp->get_rx_freq_range().start() - usrp->get_fe_tx_freq_range().start();
    double max_rx_offset =
        usrp->get_rx_freq_range().stop() - usrp->get_fe_tx_freq_range().stop();
    if (rx_offset < min_rx_offset or rx_offset > max_rx_offset) {
        std::cerr << "rx_offset must be between " << min_rx_offset << " and "
                  << max_rx_offset << " for this daughter board" << std::endl;
        return EXIT_FAILURE;
    }

    std::cout << boost::format("Calibration frequency range: %d MHz -> %d MHz")
                     % (freq_start / 1e6) % (freq_stop / 1e6)
              << std::endl;

    // set RX gain
    usrp->set_rx_gain(0);

    size_t tx_error_count = 0;
    for (double tx_lo_i = freq_start; tx_lo_i <= freq_stop; tx_lo_i += freq_step) {
        const double tx_lo = tune_rx_and_tx(usrp, tx_lo_i, rx_offset);

        // frequency constants for this tune event
        const double actual_rx_rate = usrp->get_rx_rate();
        const double actual_tx_freq = usrp->get_tx_freq();
        const double actual_rx_freq = usrp->get_rx_freq();
        const double bb_dc_freq     = actual_tx_freq - actual_rx_freq;

        // reset TX DC offset
        usrp->set_tx_dc_offset(std::complex<double>(0, 0));

        // capture initial uncorrected value
        capture_samples(usrp, rx_stream, buff, nsamps);
        const double initial_dc_dbrms =
            compute_tone_dbrms(buff, bb_dc_freq / actual_rx_rate);

        // bounds and results from searching
        double i_corr_start  = -1.0;
        double i_corr_stop   = 1.0;
        double i_corr_step   = (i_corr_stop - i_corr_start) / (num_search_steps + 1);
        double q_corr_start  = -1.0;
        double q_corr_stop   = 1.0;
        double q_corr_step   = (q_corr_stop - q_corr_start) / (num_search_steps + 1);
        double best_dc_dbrms = initial_dc_dbrms;
        double best_i_corr   = 0;
        double best_q_corr   = 0;
        while (i_corr_step >= precision or q_corr_step >= precision) {
            for (double i_corr = i_corr_start + i_corr_step;
                 i_corr <= i_corr_stop - i_corr_step;
                 i_corr += i_corr_step) {
                for (double q_corr = q_corr_start + q_corr_step;
                     q_corr <= q_corr_stop - q_corr_step;
                     q_corr += q_corr_step) {
                    const std::complex<double> correction(i_corr, q_corr);
                    usrp->set_tx_dc_offset(correction);

                    // receive some samples
                    capture_samples(usrp, rx_stream, buff, nsamps);
                    // check for TX errors in the current captured iteration
                    if (has_tx_error(tx_stream)) {
                        std::cout << "[WARNING] TX error detected! "
                                  << "Repeating current iteration" << std::endl;
                        // Undo the Q correction step
                        q_corr -= q_corr_step;
                        tx_error_count++;
                        if (tx_error_count >= MAX_NUM_TX_ERRORS) {
                            throw uhd::runtime_error(
                                "Too many TX errors. Aborting calibration.");
                        }
                        continue;
                    }
                    const double dc_dbrms =
                        compute_tone_dbrms(buff, bb_dc_freq / actual_rx_rate);

                    if (dc_dbrms < best_dc_dbrms) {
                        best_dc_dbrms = dc_dbrms;
                        best_i_corr   = i_corr;
                        best_q_corr   = q_corr;
                    }
                }
            }

            i_corr_start = best_i_corr - i_corr_step;
            i_corr_stop  = best_i_corr + i_corr_step;
            i_corr_step  = (i_corr_stop - i_corr_start) / (num_search_steps + 1);
            q_corr_start = best_q_corr - q_corr_step;
            q_corr_stop  = best_q_corr + q_corr_step;
            q_corr_step  = (q_corr_stop - q_corr_start) / (num_search_steps + 1);
        }

        if (best_dc_dbrms < initial_dc_dbrms) // keep result
        {
            result_t result;
            result.freq      = tx_lo;
            result.real_corr = best_i_corr;
            result.imag_corr = best_q_corr;
            result.best      = best_dc_dbrms;
            result.delta     = initial_dc_dbrms - best_dc_dbrms;
            results.push_back(result);
            if (vm.count("verbose"))
                std::cout << boost::format(
                                 "TX DC: %f MHz: lowest offset %f dB, corrected %f dB")
                                 % (tx_lo / 1e6) % result.best % result.delta
                          << std::endl;
            else
                std::cout << "." << std::flush;
        }

        // Reset underrun counts, start a new counter for the next frequency
        tx_error_count = 0;
    } // end for each frequency loop

    std::cout << std::endl;

    // stop the transmitter
    transmit.clear();
    transmitter.join();

    store_results(results, "TX", "tx", "dc", serial);

    return EXIT_SUCCESS;
}
